### PR TITLE
feature: re-ogranize swap state and utilize new Eco Router library

### PR DESCRIPTION
**Pull request by adamazad**
# Summary

Uniswap and its forks compute trades synchronously in the dapp. To compute a trade, the SDK requires a list of `Pair` which has both sell and buy tokens. Then compares the shortest and cheapest route. `Pair`s are fetched and stored in the Redux store. Information such as reserves is regularly updated using RPC calls to get up-to-date market state.

With Curve added to the mix, Eco Router has to do an RPC to determine estimated output from Curve pools, and then bundles all trades from Uniswap V2 forks and Curve and then compares them (see below diagram).

image

Notice that Uniswap has information available to it to compute a trade, but both Curve and COW have to wait for a remote resource to respond. Because of this, the UI needs to debounce user input in order to get a final input. For instance, the user presses key 1 to input 1 and then 0, 0, and another 0. The Eco Router has to wait few hundred milliseconds until it gets the final output of `1000` from the user, and then proceeds to use that as `sellTokenAmount`. 

## Eco Router Library 

To make the current "Eco Router" independent of the application state and hooks, we created a new library that is run in isolation. This makes Eco Router portable, and can be used in non-React environments such as Node. The two new hooks provide entry points for Swapbox to tap into Eco Router sources and fetch best trade, but there are low-level API that can be utilized as well.

### Eco Router References

**API**
- `getExactIn`
- `getExactOut`
- `sortByExecutationPrice`

**Hooks for React**

- `useEcoRouterExactIn`
- `useEcoRouterExactOut`

The library will be packaged and pushed under `@swapr/eco-router` package soon.

### CODE REVIEW

File: src/components/CurrencyInputPanel/index.tsx [hunk 1]
```diff
@@ -134,8 +136,27 @@ export default function CurrencyInputPanel({
...
-  const { t } = useTranslation()
+  // Take initial value from the parent component
+  const [localValue, setLocalValue] = useState(value)
+
+  useEffect(() => {
+    const timeoutUpdate = setTimeout(() => {
+      console.log({
+        value,
+        localValue
+      })
```
Comment (berteotti) [inline src/components/CurrencyInputPanel/index.tsx:147]: This can be deleted

**Comment by 0xVenky** [comment]
On a wallet that is connected and with Arb1 network, have both Token A and Token B filled with ETH and USDC in the swap box. Once the price is populated,change network to Gnosis chain, you see a blank screen with invariant error.

image

**Comment by adamazad** [comment]
> On a wallet that is connected and with Arb1 network, have both Token A and Token B filled with ETH and USDC in the swap box. Once the price is populated,change network to Gnosis chain, you see a blank screen with invariant error.
> 
> image

Been debugging this for a while, and so far the issue comes down to 

https://github.com/levelkdev/swapr-dapp/blob/030c516f70da487c200bb207219eeca4b94beef1/src/hooks/useUSDValue.ts#L18

We don't need to use EcoRouter for computing USD value here. I will refactor this out.

**Comment by niemam29** [comment]
Hey!
Noticed that "insufficient liquidity" message is gone and instead of it there is "swap with"
cach2

**Comment by MilanVojnovic95** [comment]
It takes some time to change the data after deleting or writing the amount ...
For example if you type 1000 and delete this you get this until the data is updated

image

**Review by denciu — APPROVED** [review]
Nice and clean!

**Comment by niemam29** [comment]
Retested and now i think that there are some problems with arbitrum - take a look at the rec:

https://user-images.githubusercontent.com/30408272/161520986-2cc65307-aeae-4664-a552-f24d4522b249.mp4

Also there is still no insufficient liquidity message

**Comment by MilanVojnovic95** [comment]
Retested and there is still the same problem on Arbitrum mentioned above by @niemam29 .
Also fixed a bug that I mentioned and commented on earlier.  ✔️

**Comment by niemam29** [comment]
Arbitrum is now working for me but there is still no insufficient liquidity button. Also noticed that Rinkeby testnet have the same problem as Arbitrum before

https://user-images.githubusercontent.com/30408272/161543013-a6668e8a-20c8-400c-9e98-79595fa4ff42.mp4

**Pull request merged**